File: packages/kotlin/uarp-sdk/src/main/kotlin/ai/snaga/uarp/Sse.kt
package ai.snaga.uarp

import java.io.IOException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOn
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.isActive
import kotlinx.coroutines.withTimeoutOrNull
import kotlinx.serialization.json.JsonElement
import okio.BufferedSource
import kotlin.math.pow
import kotlin.random.Random

/** One decoded `text/event-stream` frame. */
public data class ServerEvent(
    /** `id:` field (or the `event_id` carried inside a JSON payload), replayed
     *  as `Last-Event-ID` when the stream reconnects. */
    val id: String? = null,
    /** `event:` field; defaults to `message` or, when absent, to the `type`
     *  field inside a JSON data payload. */
    val event: String = "message",
    /** Concatenated `data:` lines, without the trailing newline. */
    val data: String = "",
    /** `retry:` field in milliseconds. */
    val retry: Long? = null,
) {
    /** Parse [data] as JSON. */
    public fun json(): JsonElement = uarpJson.parseToJsonElement(data)

    /** Decode [data] into [T]. */
    public inline fun <reified T> decode(): T =
        try {
            uarpJson.decodeFromString<T>(data)
        } catch (error: Exception) {
            throw DecodingException("failed to decode event payload: ${error.message}", data, error)
        }
}

/** Reconnection and lifecycle behaviour for [sseFlow]. All fields default to a
 *  generic, spec-compliant SSE stream; the platform-specific knobs (terminal
 *  events, an inactivity watchdog, `retry:` pacing) are opt-in so a caller that
 *  passes nothing gets standard SSE. */
public data class StreamOptions(
    /** Reconnect (replaying `Last-Event-ID`) when the stream ends. Default `true`. */
    val reconnect: Boolean = true,
    /** Reconnect attempts without progress before giving up. Default `5`. */
    val maxReconnects: Int = 5,
    /** Event names that complete the flow WITHOUT reconnecting. Empty by default:
     *  a generic stream reconnects on end and lets the caller stop it. The
     *  platform's run stream passes `done`, `run.completed`, `run.failed`,
     *  `team_run_done` (and deliberately NOT `run_done`, which the public replay
     *  route re-emits on connect). A group fan-in passes only `team_run_done`. */
    val terminalEvents: Set<String> = emptySet(),
    /** Max silence between lines before the socket is presumed dead and a
     *  reconnect is attempted. `null` disables the watchdog (a read timeout, or
     *  EOF, owns liveness instead). The platform sets this to 300 s: collapsing
     *  it with EOF made a silently-dead socket look like a finished stream and
     *  the chat went permanently quiet. */
    val inactivityTimeoutMillis: Long? = null,
    /** Base reconnect interval in ms; a `retry:` field overrides it per stream. */
    val baseRetryMillis: Long = 2_000,
    /** Cap on the reconnect backoff. */
    val maxBackoffMillis: Long = 8_000,
    /** Reconnect budget resets after this long connected without a disconnect,
     *  so a long healthy stream doesn't carry "this is the Nth retry" baggage. */
    val stabilityResetMillis: Long = 60_000,
    /** Optional connection-lifecycle observer. Fired on the same dispatcher the
     *  flow runs on (IO by default); the callback must be thread-safe. `Disconnected`
     *  is NOT fired when the caller cancels the flow — only on a natural end. */
    val onState: ((StreamState) -> Unit)? = null,
)

/** Connection-lifecycle states reported by [sseFlow] via [StreamOptions.onState]. */
public sealed interface StreamState {
    /** About to open (or reopen) the HTTP connection. Fired once, before the first attempt. */
    public data object Connecting : StreamState
    /** The server answered 200 and the stream is being read. */
    public data object Connected : StreamState
    /** Waiting on backoff before a reconnect attempt. [attempt] is 1-based. */
    public data class Reconnecting(val attempt: Int) : StreamState
    /** The flow ended without the caller cancelling it (terminal frame, `[DONE]`, or the reconnect budget exhausted). */
    public data object Disconnected : StreamState
}

/**
 * A cold [Flow] of server-sent events.
 *
 * ```kotlin
 * client.runs.streamRunEvents(runId).collect { event ->
 *     if (event.event == "run.completed") throw CancellationException()
 * }
 * ```
 *
 * Cancelling the collector closes the HTTP connection.
 *
 * The decoder handles the three wire shapes the platform emits: standard
 * `text/event-stream` frames (`event:`/`id:`/`data:`/`retry:`, blank-line
 * dispatch), a JSON object carried in an SSE comment
 * (`:{"type":"…","event_id":"…"}`), and a bare NDJSON line
 * (`{"type":"…","event_id":"…"}`). A `data: [DONE]` frame is a hard terminal:
 * any pending event is flushed and the flow completes without reconnecting.
 */
public fun sseFlow(client: UarpClient, spec: RequestSpec): Flow<ServerEvent> = flow {
    val options = spec.options.stream
    var lastEventId: String? = null
    var attempt = 0
    var baseRetry = options.baseRetryMillis

    options.onState?.invoke(StreamState.Connecting)

    reconnect@ while (currentCoroutineContext().isActive) {
        if (attempt > 0) {
            options.onState?.invoke(StreamState.Reconnecting(attempt))
            delay(streamBackoff(attempt, baseRetry, options.maxBackoffMillis))
            if (!currentCoroutineContext().isActive) break@reconnect
        }

        // On reconnect, replace any spec-supplied `Last-Event-ID` with the id
        // the last delivered event carried, so the stream resumes from there.
        val resumed = lastEventId
        val headers = buildList {
            for ((name, value) in spec.headers) {
                if (resumed != null && name.equals("Last-Event-ID", ignoreCase = true)) continue
                add(name to value)
            }
            resumed?.let { add("Last-Event-ID" to it) }
        }
        val query = if (client.sseTokenInQuery) spec.query + ("token" to client.apiKey) else spec.query
        // Streams are long-lived: the unary call timeout would cut them short.
        val attemptSpec = spec.copy(
            headers = headers,
            query = query,
            options = spec.options.copy(timeoutMillis = 0),
        )

        val request = client.buildRequest(attemptSpec, idempotencyKey = null, accept = "text/event-stream")
        val response = try {
            client.httpFor(attemptSpec.options).newCall(request).await()
        } catch (error: IOException) {
            if (!options.reconnect || attempt >= options.maxReconnects) {
                throw ConnectionException(error.message ?: "could not open event stream", error)
            }
            attempt++
            continue@reconnect
        }

        if (!response.isSuccessful) {
            val body = response.use { it.body?.string().orEmpty() }
            val problem = ApiException(response.code, parseProblem(body), response.headers.toMap(), body.ifBlank { null })
            // 401 always surfaces so the caller can act on it (the app treats it
            // as "stop the stream"); any other HTTP error retries like a dropped
            // connection while the reconnect budget lasts, then surfaces.
            if (response.code == 401 || !options.reconnect || attempt >= options.maxReconnects) throw problem
            attempt++
            continue@reconnect
        }

        options.onState?.invoke(StreamState.Connected)

        // A connection that delivered at least one event counts as progress and
        // resets the reconnect budget; one that closed immediately does not, so
        // a flapping server cannot spin this loop.
        var delivered = false
        var terminal = false
        response.use { open ->
            val source = open.body?.source() ?: throw StreamException("event stream response has no body")
            val parser = SseParser()
            val connectedAt = System.currentTimeMillis()

            read@ while (currentCoroutineContext().isActive) {
                val line = if (options.inactivityTimeoutMillis != null) {
                    withTimeoutOrNull(options.inactivityTimeoutMillis) {
                        runCatching { source.readUtf8Line() }.getOrNull()
                    }
                } else {
                    runCatching { source.readUtf8Line() }.getOrNull()
                }

                if (line == null && !source.isOpen) break@read
                if (line == null) {
                    val eof = runCatching { source.exhausted() }.getOrDefault(true)
                    if (eof) break@read
                    // Inactivity watchdog: the socket is silent but not closed.
                    // Reconnect with `Last-Event-ID` rather than treating the
                    // silence as a finished stream.
                    if (!options.reconnect || attempt >= options.maxReconnects) break@read
                    attempt++
                    continue@reconnect
                }

                // A healthy connection that survived the stability window
                // shouldn't carry "this is the Nth retry" baggage into its next
                // disconnect.
                if (attempt > 0 &&
                    System.currentTimeMillis() - connectedAt >= options.stabilityResetMillis
                ) {
                    attempt = 0
                }

                val event = parser.feed(line)
                // `data: [DONE]` may return a flushed pending event OR null,
                // but either way it sets `isDone` — so check it here, before the
                // `?: continue` would skip the terminal test on a null return.
                if (parser.isDone) {
                    if (event != null) {
                        if (event.id != null) lastEventId = event.id
                        event.retry?.takeIf { it > 0 }?.let { baseRetry = it }
                        delivered = true
                        emit(event)
                    }
                    terminal = true
                    break@read
                }
                val dispatched = event ?: continue@read
                if (dispatched.id != null) lastEventId = dispatched.id
                dispatched.retry?.takeIf { it > 0 }?.let { baseRetry = it }
                delivered = true
                emit(dispatched)
                if (dispatched.event in options.terminalEvents) {
                    terminal = true
                    break@read
                }
            }

            parser.finish()?.let { event ->
                if (event.id != null) lastEventId = event.id
                delivered = true
                if (event.event in options.terminalEvents || parser.isDone) terminal = true
                emit(event)
            }
        }

        if (!currentCoroutineContext().isActive) break@reconnect
        if (terminal) break@reconnect

        // A clean EOF without a terminal frame is a proxy/socket drop mid-run,
        // not a finished stream — reconnect with `Last-Event-ID`.
        if (delivered) attempt = 0
        if (!options.reconnect || attempt >= options.maxReconnects) break@reconnect
        attempt++
    }

    // Only a natural end reports Disconnected — a caller that cancelled the
    // flow has already decided it is done.
    if (currentCoroutineContext().isActive) options.onState?.invoke(StreamState.Disconnected)
}.flowOn(Dispatchers.IO)

/** Half-deterministic, half-random backoff: `maxSleep/2 + rand(0..maxSleep/2)`,
 *  so it climbs with attempts but clients don't all wake on the same boundary. */
internal fun streamBackoff(attempt: Int, baseIntervalMillis: Long, maxDelayMillis: Long): Long {
    val exponential = baseIntervalMillis * 2.0.pow((attempt - 1).coerceAtLeast(0))
    val maxSleep = minOf(maxDelayMillis.toDouble(), exponential)
    val half = (maxSleep / 2).coerceAtLeast(1.0)
    return (half + Random.nextDouble(0.0, half)).toLong().coerceAtLeast(0)
}

/**
 * Line-oriented `text/event-stream` decoder. Handles the three wire shapes the
 * platform emits and the `[DONE]` hard terminal; a frame with no `data:` is not
 * a deliverable event (an `id:`-only frame only updates the last event id).
 */
internal class SseParser {
    private val data = mutableListOf<String>()
    private var event = ""
    private var id: String? = null
    private var retry: Long? = null
    private var hasFields = false
    private var done = false

    /** `true` once a `data: [DONE]` frame arrived — the flow terminates without
     *  reconnecting. */
    val isDone: Boolean get() = done

    /** Feed one line; returns an event when a frame is complete. */
    fun feed(line: String): ServerEvent? {
        if (line.isEmpty()) return dispatch()

        // SSE comment. The platform also carries a JSON payload in a comment
        // (`:{"type":"…","event_id":"…"}`); that is a self-contained frame.
        // A bare comment is a keep-alive.
        if (line.startsWith(":")) {
            val body = line.drop(1).trim()
            if (body.startsWith("{")) return inlineEvent(body)
            return null
        }

        // Bare NDJSON line — a self-contained frame with no field prefix.
        if (line.startsWith("{")) return inlineEvent(line)

        val separator = line.indexOf(':')
        val field = if (separator == -1) line else line.substring(0, separator)
        val value = if (separator == -1) "" else line.substring(separator + 1).removePrefix(" ")

        hasFields = true
        when (field) {
            "event" -> event = value
            "data" -> {
                if (value == "[DONE]") {
                    done = true
                    // Flush a pending event, if any; `[DONE]` itself carries no
                    // payload.
                    return if (data.isNotEmpty() || event.isNotEmpty() || id != null) dispatch() else null
                }
                data += value
            }
            "id" -> if (!value.contains('\u0000')) id = value
            "retry" -> retry = value.toLongOrNull()
            else -> Unit // unknown fields are ignored
        }
        return null
    }

    /** Flush a frame left unterminated when the connection closed. */
    fun finish(): ServerEvent? = dispatch()

    private fun dispatch(): ServerEvent? {
        if (!hasFields) return null
        val joined = data.joinToString("\n")
        // A frame with no data is not a deliverable event: an `id:`/`retry:`-only
        // frame updates state but carries nothing to emit.
        if (joined.isEmpty()) {
            reset()
            return null
        }
        val resolved = event.ifEmpty { extractEventType(joined) } ?: "message"
        val result = ServerEvent(id = id, event = resolved, data = joined, retry = retry)
        reset()
        return result
    }

    /** A comment-JSON or NDJSON frame: type and id live inside the JSON body. */
    private fun inlineEvent(body: String): ServerEvent =
        ServerEvent(
            id = extractField(body, "event_id"),
            event = extractEventType(body) ?: "message",
            data = body,
        )

    private fun reset() {
        data.clear()
        event = ""
        id = null
        retry = null
        hasFields = false
        // `id` is per-frame: the platform's client resets it on dispatch, so an
        // event's id is only the `id:` its own frame carried (or the `event_id`
        // inside a JSON payload). The flow captures the emitted id for replay
        // before this runs, so reconnect still resumes from the last event id.
    }
}

/**
 * Pull one string field out of a JSON body WITHOUT fully decoding it — the
 * stream carries thousands of frames a minute, and a full parse per frame to
 * learn its `type` is the difference between a smooth stream and a stuttering
 * one. Honours escaped quotes so a `"` inside a value can't fool it.
 */
internal fun extractField(json: String, field: String): String? {
    val needle = "\"$field\""
    val start = json.indexOf(needle)
    if (start < 0) return null
    var i = start + needle.length
    while (i < json.length && (json[i] == ':' || json[i] == ' ')) i++
    if (i >= json.length || json[i] != '"') return null
    i++
    val valueStart = i
    while (i < json.length) {
        when {
            json[i] == '\\' -> {
                i++
                if (i >= json.length) break
            }
            json[i] == '"' -> break
        }
        i++
    }
    if (i <= valueStart) return null
    return json.substring(valueStart, i)
}

/** The `type` field of a JSON frame, peeked without decoding. */
internal fun extractEventType(json: String): String? = extractField(json, "type")

File: packages/typescript/src/core/sse.ts
/** Server-sent events: frame parsing, reconnection and an async-iterable stream. */
import { APIConnectionError, APIError } from './errors.js';
import type { JsonValue } from './json.js';
import { sleep, streamBackoffDelay } from './util.js';

export interface UarpEvent {
  /** `id:` field (or the `event_id` inside an inline JSON frame); replayed as `Last-Event-ID` when reconnecting. */
  id?: string;
  /** `event:` field, the `type` inside a JSON payload, or `message`. */
  event: string;
  /** Raw `data:` payload with the trailing newline removed. */
  data: string;
  /** `retry:` field in milliseconds, when the server sent one. */
  retry?: number;
  /** Parse `data` as JSON. Throws `SyntaxError` for non-JSON frames. */
  json<T = JsonValue>(): T;
}

/** Connection-lifecycle states reported by `EventStream` via `EventStreamOptions.onState`. */
export type StreamState =
  | { type: 'connecting' }
  | { type: 'connected' }
  | { type: 'reconnecting'; attempt: number }
  | { type: 'disconnected' };

/**
 * Side-channel from `parseEventStream` to `EventStream`: the decoder sets
 * `isDone` when it sees `data: [DONE]`, so the reconnect loop can tell a hard
 * terminal (no reconnect) from a clean EOF (reconnect). `for await…of` cannot
 * read a generator's return value, so the loop reads this instead.
 */
export class DecoderState {
  isDone = false;
  /** Mark the stream as terminated by `data: [DONE]`. */
  markDone(): void {
    this.isDone = true;
  }
}

export interface EventStreamOptions {
  signal?: AbortSignal;
  /** Reconnect with `Last-Event-ID` when the stream drops. Default `true`. */
  reconnect?: boolean;
  /** Consecutive reconnect attempts before giving up. Default `5`. */
  maxReconnects?: number;
  /** Event names that complete the stream WITHOUT reconnecting. Empty by default. */
  terminalEvents?: Set<string>;
  /** Max silence between reads before the socket is presumed dead and a reconnect
   *  is attempted. `undefined` disables the watchdog (EOF owns liveness). */
  inactivityTimeoutMillis?: number;
  /** Base reconnect interval in ms; a `retry:` field overrides it per stream. */
  baseRetryMillis?: number;
  /** Cap on the reconnect backoff. */
  maxBackoffMillis?: number;
  /** Reconnect budget resets after this long connected without a disconnect. */
  stabilityResetMillis?: number;
  /** Optional connection-lifecycle observer. `disconnected` is NOT fired when the
   *  caller aborts the stream — only on a natural end. */
  onState?: (state: StreamState) => void;
  /** Injectable clock/random for tests; default `Date.now`/`Math.random`. */
  now?: () => number;
  random?: () => number;
}

/** Opens the underlying HTTP response; called again for every reconnect. */
export type StreamConnector = (lastEventId: string | undefined, signal: AbortSignal) => Promise<Response>;

/**
 * A live SSE stream.
 *
 * ```ts
 * const stream = client.runs.streamRunEvents(runId);
 * for await (const event of stream) {
 *   if (event.event === 'run.completed') break;
 * }
 * ```
 *
 * Breaking out of the loop (or calling `close()`) aborts the HTTP request.
 */
export class EventStream implements AsyncIterable<UarpEvent> {
  readonly #connect: StreamConnector;
  readonly #controller = new AbortController();
  readonly #reconnect: boolean;
  readonly #maxReconnects: number;
  readonly #terminalEvents: Set<string>;
  readonly #inactivityTimeout: number | undefined;
  readonly #baseRetry: number;
  readonly #maxBackoff: number;
  readonly #stabilityReset: number;
  readonly #onState: ((state: StreamState) => void) | undefined;
  readonly #now: () => number;
  readonly #random: () => number;
  #lastEventId: string | undefined;
  #consumed = false;

  constructor(connect: StreamConnector, options: EventStreamOptions = {}) {
    this.#connect = connect;
    this.#reconnect = options.reconnect ?? true;
    this.#maxReconnects = options.maxReconnects ?? 5;
    this.#terminalEvents = options.terminalEvents ?? new Set();
    this.#inactivityTimeout = options.inactivityTimeoutMillis;
    this.#baseRetry = options.baseRetryMillis ?? 2_000;
    this.#maxBackoff = options.maxBackoffMillis ?? 8_000;
    this.#stabilityReset = options.stabilityResetMillis ?? 60_000;
    this.#onState = options.onState;
    this.#now = options.now ?? Date.now;
    this.#random = options.random ?? Math.random;
    if (options.signal) {
      const abort = () => this.#controller.abort(options.signal!.reason);
      if (options.signal.aborted) abort();
      else options.signal.addEventListener('abort', abort, { once: true });
    }
  }

  /** Abort the connection and end iteration. */
  close(): void {
    this.#controller.abort();
  }

  get closed(): boolean {
    return this.#controller.signal.aborted;
  }

  /** Resolve once an event matching `predicate` arrives, then close the stream. */
  async until(predicate: (event: UarpEvent) => boolean): Promise<UarpEvent | undefined> {
    for await (const event of this) {
      if (predicate(event)) {
        this.close();
        return event;
      }
    }
    return undefined;
  }

  async *[Symbol.asyncIterator](): AsyncIterator<UarpEvent> {
    if (this.#consumed) throw new Error('EventStream has already been consumed');
    this.#consumed = true;

    let attempt = 0;
    let baseRetry = this.#baseRetry;
    this.#onState?.({ type: 'connecting' });

    while (!this.#controller.signal.aborted) {
      let response: Response;
      try {
        response = await this.#connect(this.#lastEventId, this.#controller.signal);
      } catch (error) {
        if (this.#controller.signal.aborted) return;
        // 401 always surfaces so the caller can act on it (the app stops the
        // stream); any other connector error retries like a dropped connection.
        if (error instanceof APIError && error.status === 401) throw error;
        if (!this.#reconnect || attempt >= this.#maxReconnects) throw error;
        this.#onState?.({ type: 'reconnecting', attempt: attempt + 1 });
        await sleep(
          streamBackoffDelay(attempt + 1, baseRetry, this.#maxBackoff, this.#random),
          this.#controller.signal,
        );
        attempt++;
        continue;
      }

      if (!response.body) {
        throw new APIConnectionError('Event stream response has no body');
      }

      this.#onState?.({ type: 'connected' });

      // A connection that delivered at least one event counts as progress and
      // resets the reconnect budget; one that closed immediately does not, so
      // a flapping server cannot spin this loop forever.
      let delivered = false;
      let terminal = false;
      let clean = false;
      const decoder = new DecoderState();
      const connectedAt = this.#now();
      try {
        for await (const event of parseEventStream(response.body, this.#controller.signal, {
          inactivityTimeoutMillis: this.#inactivityTimeout,
          state: decoder,
        })) {
          if (event.id !== undefined) this.#lastEventId = event.id;
          if (event.retry && event.retry > 0) baseRetry = event.retry;
          delivered = true;
          yield event;
          if (this.#terminalEvents.has(event.event)) {
            terminal = true;
            break;
          }
        }
        clean = true;
      } finally {
        // Breaking out of the caller's `for await` lands here.
        if (!clean) this.#controller.abort();
      }
      // `data: [DONE]` is a hard terminal: no reconnect.
      if (decoder.isDone) terminal = true;
      if (terminal) {
        this.#onState?.({ type: 'disconnected' });
        return;
      }

      // A healthy connection that survived the stability window shouldn't carry
      // "this is the Nth retry" baggage into its next disconnect.
      if (attempt > 0 && this.#now() - connectedAt >= this.#stabilityReset) attempt = 0;
      if (delivered) attempt = 0;

      if (!this.#reconnect || this.#controller.signal.aborted) {
        this.#onState?.({ type: 'disconnected' });
        return;
      }
      if (attempt >= this.#maxReconnects) {
        this.#onState?.({ type: 'disconnected' });
        return;
      }
      this.#onState?.({ type: 'reconnecting', attempt: attempt + 1 });
      await sleep(
        streamBackoffDelay(attempt + 1, baseRetry, this.#maxBackoff, this.#random),
        this.#controller.signal,
      );
      attempt++;
    }
  }
}

/** Decode an SSE byte stream into events, per the WHATWG event-stream rules. */
export async function* parseEventStream(
  body: ReadableStream<Uint8Array>,
  signal?: AbortSignal,
  options?: { inactivityTimeoutMillis?: number; state?: DecoderState },
): AsyncGenerator<UarpEvent> {
  const reader = body.getReader();
  const decoder = new TextDecoder('utf-8');
  const inactivityTimeout = options?.inactivityTimeoutMillis;
  const state = options?.state;
  let buffer = '';
  let dataLines: string[] = [];
  let eventName = '';
  let id: string | undefined;
  let retry: number | undefined;
  let hasFields = false;

  const flush = (): UarpEvent | undefined => {
    if (!hasFields) return undefined;
    hasFields = false;
    const data = dataLines.join('\n');
    const eventId = id;
    const eventResolved = eventName;
    const retryOut = retry;
    dataLines = [];
    eventName = '';
    id = undefined; // `id` is per-frame, not persisted across frames.
    retry = undefined;
    // A frame with no `data:` is not a deliverable event: an `id:`/`retry:`-only
    // frame updates state but carries nothing to emit.
    if (data.length === 0) return undefined;
    const resolved = eventResolved || extractEventType(data) || 'message';
    return makeEvent({ id: eventId, event: resolved, data, retry: retryOut });
  };

  const onAbort = () => void reader.cancel().catch(() => {});
  signal?.addEventListener('abort', onAbort, { once: true });

  try {
    for (;;) {
      // Per-read inactivity watchdog: race the read against a timer that
      // *resolves* (it does not reject or abort the caller's signal). On
      // timeout the socket is silent but not closed — release the read and let
      // the loop reconnect with `Last-Event-ID` rather than treating the
      // silence as a finished stream.
      let readResult: ReadableStreamReadResult<Uint8Array> | 'timeout';
      if (inactivityTimeout) {
        let timer: ReturnType<typeof setTimeout> | undefined;
        const timeout = new Promise<'timeout'>((resolve) => {
          timer = setTimeout(() => resolve('timeout'), inactivityTimeout);
        });
        readResult = await Promise.race([reader.read(), timeout]);
        if (timer) clearTimeout(timer);
      } else {
        readResult = await reader.read();
      }
      if (readResult === 'timeout') {
        await reader.cancel().catch(() => {});
        return;
      }
      const { done, value } = readResult;
      if (done) break;
      buffer += decoder.decode(value, { stream: true });

      let newline: number;
      while ((newline = buffer.search(/\r\n|\r|\n/)) !== -1) {
        const matched = /\r\n|\r|\n/.exec(buffer.slice(newline))![0];
        // A lone trailing '\r' may be the first half of a CRLF split across chunks.
        if (matched === '\r' && newline === buffer.length - 1) break;
        const line = buffer.slice(0, newline);
        buffer = buffer.slice(newline + matched.length);

        if (line === '') {
          const event = flush();
          if (event) yield event;
          continue;
        }
        // SSE comment. The platform also carries a JSON object in a comment
        // (`:{"type":"…","event_id":"…"}`); that is a self-contained frame.
        // A bare comment is a keep-alive.
        if (line.startsWith(':')) {
          const body = line.slice(1).trim();
          if (body.startsWith('{')) yield inlineEvent(body);
          continue;
        }
        // Bare NDJSON line — a self-contained frame with no field prefix.
        if (line.startsWith('{')) {
          yield inlineEvent(line);
          continue;
        }

        const colon = line.indexOf(':');
        const field = colon === -1 ? line : line.slice(0, colon);
        let value = colon === -1 ? '' : line.slice(colon + 1);
        if (value.startsWith(' ')) value = value.slice(1);

        hasFields = true;
        switch (field) {
          case 'event':
            eventName = value;
            break;
          case 'data':
            if (value === '[DONE]') {
              state?.markDone();
              const flushed = flush();
              if (flushed) yield flushed;
              return; // hard terminal — the loop reconnects on nothing.
            }
            dataLines.push(value);
            break;
          case 'id':
            if (!value.includes('\0')) id = value;
            break;
          case 'retry': {
            const parsed = Number.parseInt(value, 10);
            if (Number.isFinite(parsed)) retry = parsed;
            break;
          }
          default:
            break; // unknown fields are ignored
        }
      }
    }
    const trailing = flush();
    if (trailing) yield trailing;
  } finally {
    signal?.removeEventListener('abort', onAbort);
    reader.releaseLock?.();
  }
}

/** A comment-JSON or NDJSON frame: type and id live inside the JSON body. */
function inlineEvent(body: string): UarpEvent {
  return makeEvent({
    id: extractField(body, 'event_id'),
    event: extractEventType(body) || 'message',
    data: body,
  });
}

function makeEvent(fields: { id?: string; event: string; data: string; retry?: number }): UarpEvent {
  return {
    ...fields,
    json<T = JsonValue>(): T {
      return JSON.parse(fields.data) as T;
    },
  };
}

/**
 * Pull one string field out of a JSON body WITHOUT fully decoding it — the
 * stream carries thousands of frames a minute, and a full parse per frame to
 * learn its `type` is the difference between a smooth stream and a stuttering
 * one. Honours escaped quotes so a `"` inside a value can't fool it.
 */
export function extractField(json: string, field: string): string | undefined {
  const needle = `"${field}"`;
  const start = json.indexOf(needle);
  if (start < 0) return undefined;
  let i = start + needle.length;
  while (i < json.length && (json[i] === ':' || json[i] === ' ')) i++;
  if (i >= json.length || json[i] !== '"') return undefined;
  i++;
  const valueStart = i;
  while (i < json.length) {
    const c = json[i];
    if (c === '\\') {
      i += 2;
      continue;
    }
    if (c === '"') break;
    i++;
  }
  if (i <= valueStart) return undefined;
  return json.slice(valueStart, i);
}

/** The `type` field of a JSON frame, peeked without decoding. */
export function extractEventType(json: string): string | undefined {
  return extractField(json, 'type');
}